### DISCUSSION TITLE
Plugins, Protocols, and Unholy Algebras

### DISCUSSION THREAD

**Discussion by rtbs-dev** [beartype/beartype#188] [Ideas]
Running off of #187, where some limitations of `typing` are causing fundamental issues that prevent possible functionality from existing. Thought I would move the discussion here, since we started to stray off into plugin/protocol territory, which would impact a heck of a lot more than just contravariant generics. I'm going to add ideas to this as I have time, but for now, a checklist for what we might discuss: 

**Mechanisms** (that could exploit `TypeHint` for great good:tm:)
- [x] Protocols with a "blessed" `__beartype__` attribute to override checks in complex types
- [ ] global typehint registry  with weakrefs to "blessed types" and their typehints

**Impacts** (i.e. why the heck are we talking about this)
- [ ] fantastic generics and how to handle them (contravariance ftw)
- [x] use as _tags_ to speed up complex type validation
- [x] enabling of ADTs (kinda) through tagged unions, product types, and parametric typing more generally
- [ ] easily allow type-based **dispatch** a la Rust Traits or Go interfaces... I _think_. 

I will post some initial code sketches for these, I just wanted to get things started before I take a break for lunch!

**Comment by rtbs-dev** [comment]
# Protocol Blessing
We want to be able to tell beartype (and e.g. downstream plugins) what _precisely_ a thing's "type" is, and make sure that's known _relative to_ all other types. `beartype.door.TypeHint` uses some pretty amazing black magic that does this. (see [this comment over at Plum](https://github.com/wesselb/plum/issues/53#issuecomment-1274440431) for an example of DOOR rocking our collective socks off.  
The issue is that `typing` actually removes at runtime a lot of information `TypeHint`s _need_ in order to actually instantiate, as seen in #187. 

# the Idea
So the idea is that, wherever possible, a _subset_ of our stuff can be explicitly "blessed" to have it's type known statically by beartype, which let's beartype.door do all kinds of awesome! 

## example napkin code 

```python
from beartype import beartype
from beartype import typing as bt
from beartype.door import TypeHint as TH
from abc import abstractmethod
import types

CT = bt.TypeVar("ClassType")


@bt.runtime_checkable
class BearTyped(bt.Protocol, bt.Generic[CT]):
    """in the name of our unholy bearfathers, I dub thee as _typed_!"""
    @classmethod
    @abstractmethod
    def __beartype__(cls: bt.Type[CT]) -> TH | bt.Sequence[TH]:
        ...

def beartypeable(myclass: bt.Type[CT]) -> BearTyped[CT]:
    """nasty monkey-patching, do not use"""

    @classmethod
    def __beartype__(cls: bt.Type[CT]) -> TH:
        """Default janky behavior"""
        return TH(cls)

    if not hasattr(myclass, "__beartype__"):
        setattr(myclass, "__beartype__", __beartype__)  # yes?
    return myclass
    # class MoreBearable(myclass, BearTyped):  # no
    #     def __beartype__(cls)->TH:
    #         return TH(cls)

if __name__ == "__main__":

    @beartypeable
    class Fake(bt.NamedTuple):
        name: str
        kind: str
        velocity: float

    fake_data = ("n1", "bob", 2.3)
    fake: BearTyped = Fake(*fake_data)
    FakeLike = bt.Tuple[str, str, float]  # let the shennanigans commence

    print(
        "\nbaselines\n",
        TH(bt.Tuple[str, str, float]).is_bearable(fake),  # works, Tuple >= NamedTuple
        TH(Fake).is_bearable((fake_data)),  # Doesn't, NamedTuple <= Tuple
        TH(Fake) <= TH(FakeLike),  # works, Fake is a specialization of FakeLike
        "\n\nshennanigans\n",
        TH(bt.Tuple[str, str, float]) >= Fake.__beartype__(),  # should work, right?
        TH(bt.Tuple[str, str, float]) <= Fake.__beartype__(),
        TH(BearTyped).is_bearable(fake),
        fake.__beartype__() is TH(Fake),
        Fake.__beartype__() is TH(Fake),
    )
```

```shell
baselines
 True False False 

shennanigans
 False False True True True
```
## assessment
Notice how the tuple never checks as a supertype of the namedtuple? even though it's the same thing inside? That's the generics issue. In practice, I would like the __beartype__ thing to return a _sequence of typehints_ when I want it to behave like a product type. why? Check back in our tags/adt section for more! :stuck_out_tongue: :

**Reply by leycec** [reply]
> Notice how the tuple never checks as a supertype of the namedtuple? even though it's the same thing inside? That's the generics issue.

**Gah!!!!!!!!** You may have uncovered a squishy `beartype.door.TypeHint` bug. If true, we should summarily squash that bug and then smear its green viscera all over our commit history. :mans_shoe: :bug:

## The Truth: Nobody Knows

So, `typing.NamedTuple` instances are literally just `tuple` instances with a monkey-patched `__annotations__` dunder attribute that's riding shotgun while blasting *Blind Guardian*. This *sorta* implies that `typing.NamedTuple` type hints should be treated as subtypes of `typing.Tuple[...]` type hints subscripted by the same types, *maybe*. Arguably, this should just work already:

```python
>>> from beartype.door import TypeHint
>>> TypeHint(tuple[str, str, float]) >= TypeHint(Fake)
True  # <-- this currently returns "False"; this is possibly bad
```

Semantically, that is. Pragmatically, [official documentation for the `typing` module cautions us that `typing.NamedTuple` (and user-defined subclasses thereof) are *not* intended for direct use as type hints](https://docs.python.org/3/library/typing.html#other-special-directives):

> **These are not used in annotations.** They are building blocks for declaring types.

Of course, types *are* valid annotations. From this, we can conclude that `typing` authors may not know what they're talking about – or may know what they're talking about but are failing to unambiguously explain what that they're talking about.

## The Truth: Mypy Knows

So, the world-shattering question that will shape the course of future QA for millennia to come then becomes: what should `beartype.door.TypeHint` do here? Should we either:

* Continue treating `typing.NamedTuple` (and user-defined subclasses thereof) as standard types like any other like we currently do? This is what the `typing` module and the devil on my shoulder wants us to do. *Or*...
* Refactor `beartype.door.TypeHint` to treat named tuples as subhints (in the `<=` sense) of equivalent non-named tuples subscripted by the same child type hints? This is what common sense and the angel on my shoulder wants us to do.

Dare we defy `typing` standards? @beartype goes where only eagles dare.

The more important question, I suppose, is what do **static type checkers** think about all this shadow madness? Does mypy agree with the `typing` module *or* does mypy agree with common sense? Mypy authors are surprisingly pragmatic. Mypy's final answer may shock and appal us all. Whatever mypy does, we should do. Maybe. Or maybe...

*...it's time to incite the Anti-mypy Uprising of 2022!* :fire: :boom:

**Reply by rtbs-dev** [reply]
Ah cool! Ok I'll come back to these, but I think it _is_ a bug then. 

Not in general, just for this tuple case. I knew NamedTuple was a subtype (and subhint) of tuple, and we _happen_ to know that immutable containers, as a rule, are by default _covariant_ in their parameters. 

So if the parameters are subtypes, then the types should be as well. This is why we use Sequence rather than List and stuff, so we know covariance is applicable for container types at runtime! ☺️

**Reply by leycec** [reply]
> ...we *happen* to know immutable containers, as a rule, are by default covariant in their parameters.

**Yes.** Yes, of course. We know that! Surely. Clearly. I mean, who doesn't. Right? <sup>*...tugs collar fitfully*</sup>

Okay, okay. I admit. I actually had no idea. @beartype currently just ignores the Great Covariance-contravariance Divide by uniformly type-checking *everything* covariantly. Thankfully, that seems to mostly be what everyone wants anyway. The creaking silence and ominous lack of complaints justifies our shoddy laziness.

But that's a poor excuse for ostensibly doing the wrong thing. @beartype should really be stricter and more standards-compliant about this sort of thing... especially where type variables are concerned. If you see @beartype improperly type-check something covariantly where @beartype should instead be type-checking that thing invariantly or contravariantly, **please submit an issue exposing our shame for all to see.**

In the meanwhile, let's open up a [new issue disclosing the above badness for all to see](https://github.com/beartype/beartype/issues/189).

**Reply by rtbs-dev** [reply]
I mean, this is all pretty new to me tbh, just a really weirdly _engaging_ rabbit hole for some reason. Eventually I will absolutely say something stupid and being a total n00b at this will be my excuse! (Haha, take that! Not-so-smart now, am I??) 

I did go ahead and add some other "bug" examples showing generic variance on that issue, so that we can check. I _guess_ this means something like a `test_variances.py` is warranted eventually in the test suite? I can probably adapt my example code to something looking vaguely like that, presuming you decide what direction you want to take beartype variance support. :)

**Comment by rtbs-dev** [comment]
I don't feel like writing out an example of the global registry option (but someone else feel free to say what the merits would be! I don't really care, for the sake of argument), but nonetheless we either maintain `TypeHint`s as **local** or **global** _metadata_ that gets associated with a set of predetermined types. Note that I have them associated as maps from `Type[obj] -> TH|Sequence[TH]`.

Why??

## Runtime Optimization of complex type checks
This can let us do some pretty stupid stuff, pretty fast: 

### deeply nesting? why not? 
If I know a priori that some type `MyContainer` is composed of some other objects, say, `A`,`B`,`C`, AND _each of A,B,C comply with `BearTyped`, then ostensibly one of two cases have happened: 

### Either instances or soon-to-be
1. A,B,C are already instantiated. In this case, they should have already been validated (maybe we call `beartype` inside the __init__?) to where they have a valid `__beartype__` classmethod (i.e. we trust that the contract that the `BearTyped` protocol signed at the class-object level was abided by at the instance-object level... for some _good_ reason, hopefully).
2. A,B,C are __getting__ instantiated during instantiation of `MyContainer[A|B|C]`. This is the use-case I see most often, since it's the default when using immutable stuff like  NamedTuples, etc. [^1] In this case, we're going to want to, say, use `@beartype` make sure that `MyContainer` AND one of A,B,C get correctly instantiated in a way that matches their `__beartype__` secret-sauce. 

### OK, that's the setup. Now... what do we know? 
- the `MyContainer` is "correctly instantiated" if it receives something it's allowed to receive. So if it's given something who's `__beartype__` _says it's an allowed type_, then **we believe it**. Done. No questions. These aren't the droids you're looking for. Proceed.
- The `A` (or whatever) is the one it got.... ok, so that one has no parameters. So we can actually validate this one against the instantiation arguments passed to `__init__`. 
- Perhaps most importantly, if `A` succeeds, then we can already tell beartype that `MyContainer[A]` succeeded, too! Because we trust everyone's `__beartype__`, indelibly and without question. 


### There we are!
This is how we can get tagged unions a la pydantic, or fancy nested product types like stacks of `Tuple[Tuple[...],[Tuple[..., Tuple[...],...]],...]` Or whatever. If everything in the stack is blessed by `BearTyped`, then we trust everyone's `__beartype__->TypeHint` all the way down the stack until there's no more generics (or *gasp* someone didn't bring their `__beartype__` ID with them. 

## But... how? 

The way I've written this, I'm imagining a preprocessing step _inside of `@beartype`

_**gasp**_

Ok, this may be a bad idea, since that devilry is so optimized, it's _actually insane that it works_. But what I'm hoping is that for _consenting types_, the type checker can skip the need to actually validate _everything_ in a generic, opting instead to explicitly trust whatever we _told it our intent was_ until we get to the type stack's foundation, when the check comes due. 

So, in that glorious irony, we fix the stack of nominal typing with a _teensy_-but-foundational layer of structural typing to get us going. Bad, no-good, unoptimized PoC as follows: 

```python
from beartype import beartype
from beartype import typing as bt
from beartype.door import TypeHint as TH
from beartype.door import UnionTypeHint, is_bearable
from abc import abstractmethod
import types

CT = bt.TypeVar("ClassType")


@bt.runtime_checkable
class BearTyped(bt.Protocol, bt.Generic[CT]):
    @classmethod
    @abstractmethod
    def __beartype__(cls: bt.Type[CT]) -> TH | bt.Sequence[TH]:
        ...

def beartypeable(compiletime=None):
    """nasty monkey-patching, do not use"""

    def decorator(myclass: bt.Type[CT]) -> BearTyped[CT]:
        @classmethod
        def __beartype__(
            cls: bt.Type[CT], compiletime: bt.Optional[TH] = compiletime
        ) -> TH:
            """Default janky behavior"""
            if compiletime and isinstance(cls, type):
                return compiletime
            else:
                return TH(cls)

        if not hasattr(myclass, "__beartype__"):
            setattr(myclass, "__beartype__", __beartype__)
        return myclass

    return decorator

def wrapped_is_bearable(  # forgive me, for I have copy-pasted
    # Mandatory flexible parameters.
    obj: object,
    hint,
):
    print("\n--CALLING WRAPPER--\n\n")
    if TH(hint) <= TH(BearTyped):
        myhint = hint.__beartype__()
        print(f"my hint says {hint}, but really that means a: \n\t{myhint}")
    else:
        myhint = TH(hint)
    if isinstance(obj, BearTyped):
        # forgive the use of isinstance
        print(f"\n\n I see it now!\n This hint is no mere {type(obj)}")
        mybeartype = obj.__beartype__()
        print(f"but a blessed {mybeartype}!\n My third eye has opened!\n\n")

        # oh look more isinstance
        if isinstance(myhint, UnionTypeHint):  # optimizeable overhead
            print(f"checking divergent timelines...{mybeartype.hint}")
            if mybeartype in myhint:  # verbatim matches one branch
                return True
            else:
                possibles = [wrapped_is_bearable(obj, t.hint) for t in myhint]
                # going recursive, see if any won
                return any(possibles)

        # so this is an ugly hack approximating prod type
        if myhint <= TH(tuple):
            print(f"checking convergent timelines...{mybeartype.hint}")
            if mybeartype <= TH(tuple):  # really should be...
                # well, except multiple inheritance
                # ok, and protocol overlapping... 0_o

                requirements = [
                    wrapped_is_bearable(x.hint, t.hint)
                    for x, t in zip(mybeartype, myhint)
                ]
            else:
                requirements = [
                    wrapped_is_bearable(mybeartype.hint, t.hint) for t in myhint
                ]
            print(
                "\nthese succeeded:\n\t ", list(zip(mybeartype, myhint, requirements))
            )
            return all(requirements)
        else:
            print("currently on the one true timeline")
            return myhint >= mybeartype

    else:
        print(f"Oh, this \n\t{obj}\n is a boring _normal_ thing...")
        return myhint.is_bearable(obj)


if __name__ == "__main__":
    
    FakeLike = bt.Tuple[str, str, float]  # let the shennanigans commence

    @beartypeable(compiletime=TH(FakeLike))
    class Faker(bt.NamedTuple):
        name: str
        kind: str
        velocity: float


    FakeLike = bt.Tuple[str, str, float]  # let the shennanigans commence

    # this kinda __beartype__ expansion thing has to be automated...
    @beartypeable(compiletime=TH(bt.Tuple[int, Faker.__beartype__().hint]))
    class FakeEst(bt.NamedTuple):
        id: int
        fakestuff: Faker
```
then:  
```python
    print(wrapped_is_bearable((1, ("hi", "bob", 2.8)), FakeEst))
```
```shell
❯ python -m grabble.beartypeable

--CALLING WRAPPER--


my hint says <class '__main__.FakeEst'>, but really that means a: 
        TypeHint(tuple[int, tuple[str, str, float]])
Oh, this 
        (1, ('hi', 'bob', 2.8))
 is a boring _normal_ thing...
True
```

```python
    print(
        wrapped_is_bearable(
            FakeEst(1, Faker("hi", "bob", 2.8)), bt.Tuple[int, FakeLike]
        )
    )
```

```shell
❯ python -m grabble.beartypeable

--CALLING WRAPPER--




 I see it now!
 This hint is no mere <class '__main__.FakeEst'>
but a blessed TypeHint(tuple[int, tuple[str, str, float]])!
 My third eye has opened!


checking convergent timelines...tuple[int, tuple[str, str, float]]

--CALLING WRAPPER--


Oh, this 
        <class 'int'>
 is a boring _normal_ thing...

--CALLING WRAPPER--


Oh, this 
        tuple[str, str, float]
 is a boring _normal_ thing...

these succeeded:
          [(TypeHint(<class 'int'>), TypeHint(<class 'int'>), False), (TypeHint(tuple[str, str, float]), TypeHint(tuple[str, str, float]), False)]
False
```

EDIT: did some more work to make things... more sensible. by a slim margin. Still, the second test is failing, since for some reason those equivalent types are _not_ compatible :thinking: Time for sleep now though.

**Comment by leycec** [comment]
**ZOMG.** My piddly brain succumbs to your galactic brain. I *barely* grok the forest that you're boldly clearing a path through. So let's instead hyper-focus on the trees, which is my only forte on GitHub.

I do trees like nobody's business. It's pretty much all I do, really. Thank the *Asperger's Syndrome* for small favours, everybody.

> ...global registry option...

Let's... *not* do global if we can help it. Please tell me we can help it, because I barely understand what's going on above.

Local is the way in all things except video games and @beartype, which are both admittedly impossible to do local. I live in the mouldy woods mostly inhabited by elderly retirees and rabid wild turkeys. What would locally sourced video games even look like here!?!?

> **Runtime Optimization of complex type checks**

I like what I'm hearing.

> This can let us do some pretty stupid stuff, pretty fast:

I *love* what I'm hearing. Gimme that pretty stupidly fast stuff, stat!

> `A,B,C` are getting instantiated during instantiation of `MyContainer[A|B|C]`. This is the use-case I see most often, since it's the default when using immutable stuff like `NamedTuples`, etc.

**My head is already blowing up.** So, let's pretend this is the only use case. Immutability *is* the way. Functional languages and JAX tell us this. Some people are even listening. I might be one of them.

> The `MyContainer` is "correctly instantiated" if it receives something it's allowed to receive. So if it's given something whose `__beartype__` says it's an allowed type, then we believe it.

**Okay.** This is fascinating. Flame-breathing dragons to my left; many-headed hydras to my right.

First, let's admit that @beartype (probably) *cannot* internally call a pure-Python callable resembling `wrapped_is_bearable()` in its type-checking code path. For micro-optimization purposes and those blessedly sweet near-real-time speeds, @beartype **does *not* internally call `is_bearable()` or `die_unless_bearable()` anywhere in type-checking code paths.** `is_bearable()` and `die_unless_bearable()` are *only* intended to be called by external callers; these callables are crude black boxes that publicize @beartype's stupidly micro-optimized type-checking internals to public view. 

Calling pure-Python callables incurs considerable overhead. Internally, @beartype avoids that overhead by instead inlining raw `isinstance(...)` checks in type-checking code that it dynamically generates.

In theory, that's hopefully still not an obstacle to the runtime typing revolution you're envisioning. But it kinda sounds like it might be. The usual way of augmenting raw `isinstance(...)` checks with higher-order intelligence is to define an `__instancecheck__()` dunder method in the metaclasses of the relevant classes. Sometimes you can reasonably do that... *but sometimes you can't.* Metaclass and metaclass conflicts get fugly fast. It's generally best to avoid them.

In practice, I'm unsure where that leaves us. Let's pretend we can still do everything you've outlined in a ~~stupidly~~ blessedly micro-optimized fashion. Perhaps we could:

1. Test for the existence of a `__beartype__()` dunder method on the passed object in the type-checking code that @beartype dynamically generates (e.g., with the standard `hasattr(obj, '__beartype__')` test, though `hasattr()` itself is surprisingly inefficient due to internally catching `AttributeError` exceptions internally raised by the `getattr()` builtin that it defers to).
2. If that attribute exists, defer to `is_bearable(obj, hint)`.
3. Else, defer to `isinstance(obj, hint)` as we currently do.

Superficially, that's beginning to smell disastrously inefficient. @beartype guarantees negligible <1µs type-checking overhead. Anything that comprises that is strictly *verboten*, sadly. Ludicrous speed is (pretty much) the only reason anyone uses @beartype. Yes. A great sadness is welling up in my soul.

Matters would improve considerably if we only care about `__beartype__()` methods on *type hints* rather than on both *type hints **and** arbitrary other objects being type-checked.* Since @beartype dynamically generates type-checking code at decoration time, micro-optimization is (mostly) *not* a concern at decoration time. In other words:

```python
    # This conditional is great! @beartype can do this at decoration time,
    # in which case speed is largely ignorable. Raise your hands like you
    # just don't care, everybody!
    if TH(hint) <= TH(BearTyped):
        myhint = hint.__beartype__()
        print(f"my hint says {hint}, but really that means a: \n\t{myhint}")
    else:
        myhint = TH(hint)

    # This conditional frightens me, which is less great. @beartype *CANNOT*
    # do this at decoration time but must instead perform this test at
    # *EACH* type-checking level of *EACH* nested container in *EACH*
    # type-checking call. This is compounded by the fact that `BearTyped` is
    # a protocol, which are notoriously slow to type-check. The
    # "beartype.typing.Protocol" superclass is actually an undocumented
    # macro-optimization that subclasses and overrides the notoriously slow
    # "typing.Protocol" superclass with considerably faster behaviour, but...
    # Still. It's runtime protocols. We really should *NOT* be testing every
    # single thing to be type-checked against a runtime protocol. Even
    # optimizing this into a much faster, mostly equivalent
    # "hasattr(obj, '__beartype__')" call would impose considerable
    # slo-mo across the board for general @beartype users -- who will then
    # have my GitHub avatar on a pike that they parade on their front lawn
    # and then burn in ignominy on Leycec Fawkes Day.
    if isinstance(obj, BearTyped):
```

The window of progress is closing. Gah! I can feel the fetid breath of Python inefficiency breathing down my ice-burned neck even as I type these broken words.

I'm spent now. Please tell me it's not all for naught! You've finalized such a phenomenal doctoral-level thesis on "Dynamic Product Types: The @beartype Way" here. Squandering the genius you so proudly flaunt above would be the baddest thing I've done all month (followed only by playing twelve continuous hours of *Stellaris* last Sunday night, which I'm less proud of than you might believe).

I too want to destroy Pydantic (...in the *nicest* way possible, of course) with a competing `beartype.producttypes` API. But I don't know if we can do that without destroying everything else that makes @beartype equally special and boring.

@beartype: *why you so rabidly obsessed with runtime complexity, bro.* :face_exhaling:

**Comment by leycec** [comment]
**Gah!** I thought for two real-world seconds and realized belatedly that this...

```python
     if isinstance(obj, BearTyped):
```

...probably *can* also be relegated to decoration time, maybe? Specifically, we would perform that check in the private `beartype._decor.decorcore._beartype_type()` decorator that decorates types. Here, `obj` would then be a `cls` and we would instead rewrite the above to resemble:

```python
def _beartype_type(...):
     ...
     if issubclass(cls, BearTyped):
```

Presumably, this would also necessitate refactoring the `__beartype__()` *instance* method callable on instances into a `__beartype__()` *class* method callable on classes. I stroke my double chin thoughtfully. :thinking: 

Would operating on the *class* rather than the *instance* of that class demolish your Steven Hawking-tier QA dissertation? Or is the fragile house of our hopes and dreams now thoroughly crushed beyond repair and useful commentary?

**Reply by rtbs-dev** [reply]
> Would operating on the class rather than the instance of that class demolish your Steven Hawking-tier QA dissertation? Or is the fragile house of our hopes and dreams now thoroughly crushed beyond repair and useful commentary?
 
So I actually am _super_ glad you asked this, I struggled the longest on this when I was cooking things up. _Notice the `@classmethod` in the definition for `BearTyped`? That there "tree" is a stinker. I went ahead and made all of them `classmethod`s by _default_, and even went down a horrifying rabbithole trying to understand whether I could dynamically add classmethods at runtime. (Cue an hour unraveling the bound vs unbound method distinction, only to realize it was what `types.MethodType` was for if you gave it a `None`, only to realize that it completely changed in Python 3+ to disallow such shennanigans, only to realize I had spelunked too deeply into this cavernous wasteland and said "screw it, I'm not doing that" and used a monkey patching class factory --- new band name, calling it --- the way the ancient NamedTuple gods intended.)

My thinking was that, with deeply-nested generics, I want to spend time comparing `type` objects with `TypeHint` wrappers, for as long as I can, until I am forced to handle the inner-most type parameter. If we check _that_ concrete type against an instance (a la the `type_of` approximation thing you and I are .... too familiar with by now), then the whole "stack" of generics is trusted through the class-method logic. 

This is why my proof-of-concept wrapper function is recursive in `type`, until it hits an `object` that isn't a `type`, where it calls `is_bearable`. Obviously that design is inefficient compared to something at decoration-time, but, YES classmethods all the way!

**Comment by leycec** [comment]
> "screw it, I'm not doing that" and used a monkey patching class factory --- new band name, calling it --- the way the ancient NamedTuple gods intended.

![black lagoon fights back](https://media.tenor.com/rtilmV7A9n0AAAAS/cthulhu-blacklagoon.gif)
<sup>Pictured: `types.MethodType` (left); @beartype (right).</sup>

> YES classmethods all the way!

*Yes, yes, YES!!!* I feel warm and tingly inside. Other questions raise their sinuous snake-like muzzles, however:

* Does `typing.Protocol` even check `@classmethod`-decorated methods? I. Have. No. Idea. Your brilliant code suggests it does, which is fascinating. I shrug nonchalantly.
* When you find a spare moment hiding between the couch cushions, would you mind bravely refactoring the `if isinstance(obj, BearTyped):` statement into something resembling `if issubclass(type(obj), BearTyped):` (assuming protocols support class methods) or perhaps just `if hasattr(type(obj), '__beartype__')` (if protocols do *not* support class methods)? Actually, do protocols even support `issubclass()` checks? I kinda suspect not. Curse your rotten API, `typing` module!

If that works (*...it won't*), we're in the clear for fun with dynamic code generation at @beartype decoration time. My dim suspicion is that most of the magic in the `wrapped_is_bearable()` tester – including both the `UnionTypeHint` magic *and* the product `tuple` type magic – can be profitably shifted into subclass implementations of the `beartype.door.TypeHint.is_subhint()` method. Doing so would then reduce the code that @beartype needs to dynamically generate at decoration time to just:

```python
    # If the current type hint annotating the current callable being
    # decorated by @beartype is a "beartyped thing" (i.e., the type of
    # this type hint defines a __beartyped__() dunder method), then
    # just generate type-checking code efficiently deferring to the
    # beartype.door.TypeHint.is_subhint() method.
    if is_beartyped_thing(hint):
        # "beartype.door.TypeHint" object or sequence of such objects
        # describing this type hint.
        hint_wrapper = hint.__beartype__()

        #FIXME: This is clearly insufficient. "{{pith}}" is the
        #current object being type-checked (e.g., parameter,
        #return value, something passed to is_bearable()). That
        #will pretty much *NEVER* be a "TypeHint" object. So,
        #we'll need to test "is_beartyped_thing({{pith}})" on 
        #that object as well and, if true, call
        #"{{pith}}.__beartyped__()" (*or something something*) on
        #that object before passing it to `is_superhint()` below.
        #WHO EVEN KNOWS WHAT'S GOING ON HERE ANYMORE?!?!?!?!?!?!?

        # Pretend you didn't see substrings like "{{hint_wrapper}}"
        # or "{{pith}}" here. That's just to remind us to generate
        # code operating on the expected objects at type-check time.
        return '{{hint_wrapper}}.is_bearable({{pith}})'
```

## ELI5 This Like I Don't Know What I'm Doing

*...because I kinda don't.* I think what we're doing here is defining an extensible plugin architecture enabling decisive and courageous end users to override standard `isinstance()`-driven type checking with type checking defined by said end users.

The only issue then is that said end users *probably* want some means of telling the `beartype.door.TypeHint` API what to do, right? Currently, that API hard-codes the mapping between low-level type hints and high-level `TypeHint` subclasses handling those type hints in [the private `beartype.door._doordata.get_typehint_subclass()` getter function](https://github.com/beartype/beartype/blob/8f33d55d7673dc7f2e57216ed03e27d9b971f443/beartype/door/_doordata.py#L59).

I'm suspecting we now need to either:

* **Support manual registration.** Specifically, @beartype could add a public registration function of some sort – say, a new `beartype.door.add_typehint_subclass()`. Users would then explicitly call that function with their custom `TypeHint` subclasses. That's quite a bit crude, however. This is Python. We could do better, including...
* **Support automatic registration.** `TypeHint` currently has a unique private `_TypeHintMeta` metaclass for various profane and foul reasons. We could augment that metaclass to implicitly register each concrete `TypeHint` subclass  that everybody creates (including our own). Sounds too good to be true, because it probably is.

More importantly, @beartype will now need to publicly expose **signs.** Signs are how @beartype internally categorizes PEP-compliant type hints. For example:

* The sign of `list[int]` is `HintSignList`.
* The sign of `int` is... `None`! *Hah.*

Signs are currently sequestered away in a private submodule of @beartype. We'll need to publish that. This is all starting to get intensely sweaty – but these are fascinating discussions worth discussing. Somehow, someway, @beartype *absolutely* does need sane plugin architecture.

This may be the start of something expensive but also beautiful. Let's see if anybody does anything. I can barely even release `beartype 0.12.0`! So that anybody may not be @leycec for some unspecified duration of time exceeding several months (possibly years). My turpitude curdles even my blood. :face_exhaling:

**Comment by leycec** [comment]
**Wam. Bam.** Plugin API begins with a whimper at feature request #191, where we delineate the first of many shambling steps to be taken towards a full-featured @beartype plugin API. I did this so that I wouldn't forget anything. I'm already forgetting everything, though. Stay inside my head, you half-formed and vaguely incoherent conceptual metaphors! :brain: :thought_balloon: :boom: 

We begin with **signs.** You'd think that would be mundane and trivial. I thought so, too. Then I wrote everything up. I am now listless and half-dead at 2:18AM in the morning – which only goes to show that you can't trust anything in this world.

**Comment by leycec** [comment]
**All specced out.** Please see feature request #193 for a *very* cursory draft discussion of the proposed `__beartype_door__()` protocol or #191 and #192 for related (albeit lower-level) plugin APIs. Thanks for all the investigatory legwork, @tbsexton. You do the hard work.

These long-winded late-night diatribes are for you. Who else is gonna read this stuff!?!? `</sigh>`

**Comment by rtbs-dev** [comment]
## This idea was bad, and I should feel bad

So I realized after name-dropping `phantom-types` to @langfield that The above nightmare is probably not the way. 

Metaclasses exist, and `__init_subclass__` is cool, yo. 

But seriously, I think basically all of the instance-vs-type shenanigans I was doing above are totally done cleaner by Phantom. The one issue was that they have to do a type-check the old TypeGuard way, meaning one-per-container-element (i.e. slow). So to answer your question

> ...because I kinda don't. I think what we're doing here is defining an extensible plugin architecture enabling decisive and courageous end users to override standard isinstance()-driven type checking with type checking defined by said end users.

Yes. Well, kinda. I think I wanted a way 
- to override a type check like `is_bearable` or `isinstance`, to rely on 
- a priori class-based information (like a TypeHint metadata, say)

and ... the way to _get there_, for my brain, was to provide a mechanism for users to override type checking in general. Whether they do the smart thing and rely on beartype.DOOR, or do insane, singularity-causing eldritch horrors was on them. 

Since I see you've already launched an all-out-war on things, would you mind taking a look at the horrifying fork I've started over at [Kermode Types](https://github.com/tbsexton/kermode-types/pull/1) that was meant to really dig into how phantom-types works and what beartype could to to speed it up? And mostly, _how does the new plugin system make that easier?_

**Comment by leycec** [comment]
**Glargh!** I'd forgotten all about the lucid mysteries of `__init_subclass__()` despite [relentlessly upvoting this StackOverflow answer on the topic five years ago](https://stackoverflow.com/a/45400561/2809027). I probably could and should have used `__init_subclass__()` throughout the @beartype codebase rather than desperately resorting to profane trickery with metaclasses. Instead, I did the bad thing.

Further evidence we're on the non-ideal timeline.

> ...dig into how `phantom-types` works...

**Ah-ha!** I can answer this... I think. `phantom-types` leverages:

* The [`__instancecheck__()` metaclass dunder method standardized by PEP 3119](https://peps.python.org/pep-3119), enabling `phantom-types` to define custom `isinstance()` logic. This is how `phantom-types` transparently supports runtime type-checking.
* Standard `typing` machinery like `typing.Protocol` and `typing.Generic`. This is how `phantom-types` transparently supports static type-checking.

Static type-checking is mostly a non-goal for us. I mean, it's vital that we not *break* static type-checking too badly – but @beartype generally doesn't go out of its way to support static type-checking until users begin raising pitch forks and wildly storming the castle gates.

Runtime type-checking, of course, is the goal – but there's more than five hundred billion ways to skin that QA onion. `__instancecheck__()` is one such way. It's also the slowest way, because it requires multiple calls of pure-Python callables for every single `isinstance()` call. That's slow, so... bad for @beartype purposes.

> ...what beartype could to to speed it up?

*...not do what `phantom-types` does.* `phantom-types` is great, but @beartype neither requires nor benefits from that sort of genericity. We're not here to support `typeguard`. We're here to *replace* `typeguard` with ourselves. Ergo, `__instancecheck__()` is (probably) not the way for us.

Instead, whatever way can most effectively support beartype validators is the way. Beartype validators minimize (and, in many cases, entirely eliminate) the number of pure-Python calls needed to perform a single type-check. This is why [our syntactic sugar for the proposed `__beartype_hint__()` dunder classmethod will *probably* leverage alternate trickery like `__mro_entries__()`](https://github.com/beartype/beartype/issues/192).

`__mro_entries__()` has the benefit of *not* requiring metaclasses, of happening at fast class creation time (rather than slow `isinstance()` check time), and of supporting beartype validators. I think. Let's contemplate the vacuous nature of my statements as we stare into the Nietzschean void together.

> And mostly, how does the new plugin system make that easier?

I also have no answers – only more questions. If @beartype's prospective New Plugin System of Wonders can't help us do things easier and/or faster, then it's bad and I should feel bad. :face_exhaling:
